trecs/models/mf.py

```diff
@@ -0,0 +1,294 @@
+        self.model_params = model_params
+        self.als_model = None  # initialize empty model
+        self.all_interactions = pd.DataFrame(columns=["user", "item"])  # empty interactions matrix
+
+        # generate user and item representations as needed
+        # note that these will be overwritten for all users/items with at least
+        # one interaction by a future call to `fit_mf`
+        gen = Generator(seed=seed)
+        if user_representation is None:
+            user_representation = gen.normal(size=(num_users, num_attributes))
+        if item_representation is None:
+            item_representation = gen.normal(size=(num_attributes, num_items))
+        # if the actual item representation is not specified, we assume
+        # that the recommender system's beliefs about the item attributes
+        # are the same as the "true" item attributes
+        if actual_item_representation is None:
+            actual_item_representation = np.copy(item_representation)
+
+        measurements = [MSEMeasurement()]
+
+        super().__init__(
+            user_representation,
+            item_representation,
+            actual_user_representation,
+            actual_item_representation,
+            num_users,
+            num_items,
+            num_items_per_iter,
+            probabilistic_recommendations=probabilistic_recommendations,
+            measurements=measurements,
+            verbose=verbose,
+            seed=seed,
+            **kwargs
+        )
+
+    def _update_internal_state(self, interactions):
+        """
+        At each training timestep, we keep track of the user/item interactions.
+        """
+        user_item_ids = tuple(zip(self.users.user_vector, interactions))
+        interaction_df = pd.DataFrame(user_item_ids, columns=["user", "item"])
+        self.all_interactions = self.all_interactions.append(interaction_df, ignore_index=True)
+
+    def update_predicted_scores(self):
+        """
+        Calculates predicted scores for every user-item pair by first running an
+        implicit MF algorithm on the interaction data. A glorified wrapper function,
+        but necessary to keep the functionality in BaseRecommender working.
+        """
+        # if no interactions have happened yet, do a simple dot product
+        if self.all_interactions.size == 0:
+            super().update_predicted_scores()
+        else:
+            self.fit_mf()
+
+    def run(
+        self,
+        timesteps=50,
+        startup=False,
+        train_between_steps=False,
+        random_items_per_iter=0,
+        vary_random_items_per_iter=False,
+        repeated_items=True,
+        no_new_items=False,
+    ):
+        """
+        Just a simple wrapper so that by default, the RS does not refit the ImplicitMF model
+        at every timestep of the simulation. See the `run` method in
+        :`~models.recommender.BaseRecommender` for more details about what this does.
+        """
```

Comment:
Something I don't understand: what happens if the user calls `run(train_between_steps=True)`?

Reply:
Then it will refit a MF model every single iteration based on just the interactions from that iteration.

Reply:
Can you add a warning here as we discussed?

Reply:
Yes, will do! Wasn't finished with the changes yet 😅

Reply:
Change has been made, so resolving this conversation!